[fs/http] Reload certificates when they change

salehqt opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

In corporate environments, client certificates have short life times for added security, and they get renewed automatically. This means that client certificate can expire in the middle of long running command such as `mount`.

This commit adds a simple timestamp check to automatically reload the certificate if it has been renewed on disk.

#### Was the change discussed in an issue or in the forum before?

No

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate. **No documentation needed**
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw reviewed (commented):
I think this is a great idea :-)

I'd like to avoid doing two Stat's on every call to RoundTrip. There can be 1000s of those per second.

Note also that RoundTrip can be called concurrently and the code you've written isn't safe for concurrent use.

It would be easy to add a time limit so it only checks for the new cert once per 1s/10s/60s or so.

I think adding a mutex will be enough to fix most of the concurrency problems.

A final thought - does anything need reloading when these certs are updated?

Thank you

salehqt commented:
> I think this is a great idea :-)
> 
> I'd like to avoid doing two Stat's on every call to RoundTrip. There can be 1000s of those per second.
I have a better way of avoiding the Stats: checking the expiration date of the certificate. I will code it up when I have the chance. In practice, overhead of Stat is negligible compared to HTTP requests.
> 
> Note also that RoundTrip can be called concurrently and the code you've written isn't safe for concurrent use.
> 
I agree, the code doesn't have any protection around writes, but it is idempotent. In practice doesn't cause any problems. I will add mutex just to improve reliability.

> It would be easy to add a time limit so it only checks for the new cert once per 1s/10s/60s or so.
> 
> I think adding a mutex will be enough to fix most of the concurrency problems.
> 
> A final thought - does anything need reloading when these certs are updated?
This code have been working for a while. I am only publishing it now as a part of upstreaming our fork of rclone.
> 
> Thank you

Any code examples on how to test this feature would be appreciated.

salehqt commented:
@ncw , I made the changes you requested:
 - Replaced Stat calls with parsing certificate and checking NotAfter time
 - Added mutex for concurrent reloading of certificates
 - Added tests for this feature

ncw commented:
I have rebased this and I'll merge when the CI goes green. Thanks for doing this @salehqt
